packages/react-data-grid/src/Canvas.tsx

```diff
@@ -0,0 +1,398 @@
+    }
+  };
+
+  onHitTopCanvas = () => {
+    const { current } = this.canvas;
+    if (current != null) {
+      current.scrollTop -= this.props.rowHeight - this.getClientScrollTopOffset(current);
+    }
+  };
+
+  onHitLeftCanvas = ({ idx }: Position) => {
+    this.scrollToColumn(idx);
+  };
+
+  onHitRightCanvas = ({ idx }: Position) => {
+    this.scrollToColumn(idx);
+  };
+
+  scrollToRow(scrollToRowIndex: number) {
+    const { current } = this.canvas;
+    if (!current) return;
+    const { rowHeight, rowsCount, height } = this.props;
+    current.scrollTop = Math.min(
+      scrollToRowIndex * rowHeight,
+      rowsCount * rowHeight - height
+    );
+  }
+
+  scrollToColumn(idx: number) {
+    const { current } = this.canvas;
+    if (!current) return;
+
+    const { scrollLeft, clientWidth } = current;
+    const newScrollLeft = getColumnScrollPosition(this.props.columns, idx, scrollLeft, clientWidth);
+    if (newScrollLeft !== 0) {
+      current.scrollLeft = scrollLeft + newScrollLeft;
+    }
+  }
+
+  getRows(rowOverscanStartIdx: number, rowOverscanEndIdx: number) {
+    if (Array.isArray(this.props.rowGetter)) {
+      return this.props.rowGetter.slice(rowOverscanStartIdx, rowOverscanEndIdx);
+    }
+    const rows = [];
+    let i = rowOverscanStartIdx;
+    while (i < rowOverscanEndIdx) {
+      const row = this.props.rowGetter(i);
+      let subRowDetails: SubRowDetails | undefined;
+      if (this.props.getSubRowDetails) {
+        subRowDetails = this.props.getSubRowDetails(row);
+      }
+      rows.push({ row, subRowDetails });
+      i++;
+    }
+    return rows;
+  }
+
+  getScroll() {
+    const { scrollTop, scrollLeft } = this.canvas.current!;
+    return { scrollTop, scrollLeft };
+  }
+
+  getClientScrollTopOffset(node: HTMLDivElement) {
+    const { rowHeight } = this.props;
+    const scrollVariation = node.scrollTop % rowHeight;
+    return scrollVariation > 0 ? rowHeight - scrollVariation : 0;
+  }
+
+  isRowSelected(idx: number, row: RowData) {
+    // Use selectedRows if set
+    if (this.props.selectedRows) {
+      const selectedRows = this.props.selectedRows.filter(r => {
+        const rowKeyValue = typeof row.get === 'function' ? row.get(this.props.rowKey) : row[this.props.rowKey];
+        return r[this.props.rowKey] === rowKeyValue;
+      });
+      return selectedRows.length > 0 && selectedRows[0].isSelected;
+    }
+
+    // Else use new rowSelection props
+    if (this.props.rowSelection) {
+      const { keys, indexes, isSelectedKey } = this.props.rowSelection as { [key: string]: unknown };
+      return isRowSelected(keys, indexes, isSelectedKey, row, idx);
+    }
+
+    return false;
+  }
+
+  setScrollLeft(scrollLeft: number) {
+    const { current } = this.interactionMasks;
+    if (current && current.setScrollLeft) {
+      current.setScrollLeft(scrollLeft);
+    }
+
+    this.rows.forEach((r, idx) => {
+      const row = this.getRowByRef(idx);
+      if (row && row.setScrollLeft) {
```

Comment:
In this case, it cannot be but I am not sure about `getRowTop ` and `getRowHeight`. It mostly likely cannot be falsy because as we use these methods in interaction mask which sits on a visible row. Lets leave it for now just to be on the safer side